drivers/dai/intel/alh/alh.h

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (c) 2022 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef __INTEL_DAI_DRIVER_ALH_H__
+#define __INTEL_DAI_DRIVER_ALH_H__
+
+#include <stdint.h>
+#include <drivers/dai.h>
+#include "alh_map.h"
+
+#define DAI_NUM_ALH_BI_DIR_LINKS_GROUP 4
+
+#define ALH_STREAM_OFFSET 0x4
+
+#define IPC4_ALH_MAX_NUMBER_OF_GTW 16
+#define IPC4_ALH_DAI_INDEX_OFFSET 7
+
+#define ALH_GPDMA_BURST_LENGTH 4
+
+#define SET_BITS(b_hi, b_lo, x)	\
+	(((x) & ((1ULL << ((b_hi) - (b_lo) + 1ULL)) - 1ULL)) << (b_lo))
+
+#define ALHASCTL_OSEL(x)          SET_BITS(25, 24, x)
+
+#define dai_get_drvdata(dai) &dai->priv_data
+#define dai_base(dai) dai->plat_data.base
+
+#define DAI_DIR_PLAYBACK 0
+#define DAI_DIR_CAPTURE 1
+
+#define ALH_CHANNELS_DEFAULT 2
+#define ALH_RATE_DEFAULT 48000
+#define ALH_WORD_SIZE_DEFAULT 32
+
+#if CONFIG_INTEL_ADSP_CAVS
+#define ALH_TXDA_OFFSET 0x400
+#define ALH_RXDA_OFFSET 0x500
+#else
+#define ALH_TXDA_OFFSET 0
+#define ALH_RXDA_OFFSET 0x100
+#endif
+
+union dai_intel_ipc4_gateway_attributes {
+	/**< Raw value */
+	uint32_t dw;
+
+	/**< Access to the fields */
+	struct {
+		/**< Gateway data requested in low power memory. */
+		uint32_t lp_buffer_alloc : 1;
+
+		/**< Gateway data requested in register file memory. */
+		uint32_t alloc_from_reg_file : 1;
+
+		/**< Reserved field */
+		uint32_t _rsvd : 30;
+	} bits; /**<< Bits */
+} __packed;
+
+/* ALH Configuration Request - SOF_IPC_DAI_ALH_CONFIG */
+struct dai_intel_ipc3_alh_params {
+	uint32_t reserved0;
+	uint32_t stream_id;
+	uint32_t rate;
+	uint32_t channels;
+
+	/* reserved for future use */
+	uint32_t reserved[13];
+} __packed;
```

Comment:
I'm guessing you have this struct (and others in this file) marked __packed as you're getting them over the wire from the SOF driver on the host?  That seems fragile to me.  If you really want unified config structs like that, we should arrange for some other layer to specify them.  Baking them into the driver internals (this header isn't part of the Zephyr API!) seems like the wrong place.

Also, packed is a noop for a struct containing only a single field type anyway.

Also also, this struct has a size of ... 17 words?  Not 16?  Is there an off-by-one in the computation of that reserved[] array?  Is the glitch symmetric with the SOF definition?  This kind of thing is exactly why you want protocol details and driver layer separated. :)

Reply:
For now I can see in SOF we have exactly the same config with 17 words :) Need to check that